src/daemon/absorber_test.cc

```diff
@@ -374,5 +383,148 @@ TEST_F(AbsorberTest, BadMessageStatus) {
       << "Daemon must not store references to the connection";
 }
 
+TEST_F(AbsorberTest, BadEmitterVersion) {
+  const String expected_host = "fake_host";
+  const ui16 expected_port = 12345;
+  const String compiler_version("compiler_version");
+  const String compiler_path("compiler_path");
+
+  conf.mutable_absorber()->mutable_local()->set_host(expected_host);
+  conf.mutable_absorber()->mutable_local()->set_port(expected_port);
+  conf.mutable_absorber()->set_min_emitter_version(kGoodEmitterVersion);
+  auto* version = conf.add_versions();
+  version->set_version(compiler_version);
+  version->set_path(compiler_path);
+
+  listen_callback =
+      [&expected_host, expected_port](const String& host, ui16 port, String*) {
+    EXPECT_EQ(expected_host, host);
+    EXPECT_EQ(expected_port, port);
+    return true;
+  };
+  connect_callback = [](net::TestConnection* connection) {
+    connection->CallOnSend([](const net::Connection::Message& message) {
+      EXPECT_TRUE(message.HasExtension(net::proto::Status::extension));
+      const auto& status = message.GetExtension(net::proto::Status::extension);
+      EXPECT_EQ(net::proto::Status::BAD_EMITTER_VERSION, status.code());
+
+      EXPECT_FALSE(message.HasExtension(proto::Result::extension));
+    });
+  };
+
+  absorber.reset(new Absorber(conf));
+  ASSERT_TRUE(absorber->Initialize());
+
+  auto connection1 = test_service->TriggerListen(expected_host, expected_port);
+  {
+    SharedPtr<net::TestConnection> test_connection =
+        std::static_pointer_cast<net::TestConnection>(connection1);
+
+    auto message(CreateMessage(""_l, ""_l, ""_l, ""_l, kBadEmitterVersion));
+    EXPECT_TRUE(
+        test_connection->TriggerReadAsync(std::move(message), StatusOK()));
+
+    UniqueLock lock(send_mutex);
+    ASSERT_TRUE(send_condition.wait_for(lock, std::chrono::seconds(1),
+                                        [this] { return send_count == 1; }));
+  }
+
+  auto connection2 = test_service->TriggerListen(expected_host, expected_port);
```

Comment:
This part doesn't correspond for the test's purpose - according to its name. It's better to leave only the first connection.